crucible-mir-comp/src/Mir/Compositional/Clobber.hs

```diff
@@ -36,46 +36,80 @@ import Mir.Compositional.Convert
 
 -- Helper functions for generating clobbering PointsTos
 
--- | Replace each primitive value within `rv` with a fresh symbolic variable.
-clobberSymbolic :: forall sym p t st fs tp0 rtp args ret.
-    (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, HasCallStack) =>
-    sym -> ProgramLoc -> String -> TypeShape tp0 -> RegValue sym tp0 ->
-    OverrideSim (p sym) sym MIR rtp args ret (RegValue sym tp0)
-clobberSymbolic sym loc nameStr shp0 rv0 = go shp0 rv0
+-- | Apply `f` to all of the `RegValue`s within a `RegValue`.  This calls `f`
+-- only on the immediate descendants of `rv`; it does not perform a recursive
+-- traversal.
+traverseTypeShape :: forall sym p t st fs tp0 rtp args ret.
+  (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, HasCallStack) =>
+  sym -> String ->
+  (forall tp.
+    TypeShape tp ->
+    RegValue sym tp ->
+    OverrideSim (p sym) sym MIR rtp args ret (RegValue sym tp)) ->
+  TypeShape tp0 -> RegValue sym tp0 ->
+  OverrideSim (p sym) sym MIR rtp args ret (RegValue sym tp0)
+traverseTypeShape sym nameStr f shp0 rv0 = go shp0 rv0
   where
     go :: forall tp. TypeShape tp -> RegValue sym tp ->
         OverrideSim (p sym) sym MIR rtp args ret (RegValue sym tp)
     go (UnitShape _) () = return ()
-    go shp@(PrimShape _ _btpr) _rv = freshSymbolic sym loc nameStr shp
+    go (PrimShape _ _) _ = die "PrimShape unimplemented"
```

Comment:
Do you think it would be worth stating in the Haddocks that the input `TypeShape` shouldn't be a `PrimShape`?

For what it's worth, I think we _could_ implement a `PrimShape` case (by just returning the `RegValue` changed, since a primitive value shouldn't have any other nested `RegValue`s). On the other hand, perhaps we don't want to, since the idea is that each call site should provide a separate implementation for `PrimShape`?

Reply:
I added a `PrimShape` case in 3a7389b7ecfcd1dff95fd37b3e1e08c95618b2e5 so the behavior of `traverseTypeShape` is easier to understand.  I also added a comment noting that it doesn't support certain shapes, like `EnumShape`.